### PR TITLE
Delegation

### PR DESCRIPTION
**Pull request by td202**
## Purpose

Implements delegation to bakers and a lock-up pool.

## Changes

- `ChainParameters` are parametrised by a `ChainParametersVersion`.
  + `ChainParametersV0` applies to protocol versions <= `P3`; `ChainParametersV1` applies to protocol version `P4`.
  + `MintDistribution` (and hence `RewardParmeters`) no longer includes the mint-per-slot rate as of `ChainParametersV1`.
  + The `_cpBakerExtraCooldownEpochs` parameter is abstracted by a new parametrised type `CooldownParameters`. For `ChainParametersV1` this includes the cool down duration for pool owners and the cool down duration for delegators; both of these are expressed in seconds (rather than epochs as before).
  + A new `TimeParameters` parameter set is introduced. For `ChainParametersV1` this consists of the reward period length (in epochs) and the mint rate per payday.
  + The `_cpBakerStakeThreshold` parameter is abstracted by a new parametrised type `PoolParameters`. For `ChainParametersV1` this consists of:
    - The commission rates for the L-pool.
    - The commission rate bounds for baker pools.
    - The minimum equity capital required to become a baker (formerly the baker stake threshold).
    - The baker capital bound.
    - The baker pool leverage bound.
- P4 update changes:
  + Add authorizations for cooldown and time parameters (`ChainParametersV1`)
  + Updates for level 2 keys have a new constructor for updating them in `ChainParametersV1`.
  + New `UpdatePayload`s are introduced for the parameters that are changed (`MintDistribution` and `PoolParameters`) or added (`CooldownParameters` and `TimeParameters`) in `ChainParametersV1`.
  + Deserialization for `UpdatePayload` is now parametrised by the protocol version. Note, however, that `UpdatePayload` is not itself parametrised, and under certain circumstances `P4` consensus may need to gracefully handle obsolete protocol updates.
- A utility module implementing binary search on a sorted vector.
- Testing: refactor QuickCheck generators into a `Generators` module. Where appropriate, generators are parametrised by protocol versions (or chain parameter versions).

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [ ] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: haskell-src/Concordium/Utils/BinarySearch.hs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+module Concordium.Utils.BinarySearch where
+
+import qualified Data.Vector as Vec
+
+-- |Perform a binary search on a sorted vector, returning the index and the value, if found.
+-- The first argument is a function that maps a value to its search key.
+-- The elements of the vector must be sorted in ascending order with respect to the key, with no
+-- duplicate keys.
+binarySearchI :: (Ord k) => (a -> k) -> Vec.Vector a -> k -> Maybe (Int, a)
+binarySearchI valToKey vec key = search 0 (Vec.length vec - 1)
```
Comment (abizjak) [inline haskell-src/Concordium/Utils/BinarySearch.hs:11]: It's very sad that this is necessary. There is an implementation of binary search https://hackage.haskell.org/package/vector-algorithms-0.8.0.4/docs/Data-Vector-Algorithms-Search.html Why was that not used?

If this implementation is used in performance sensitive contexts it should be benchmarked.

File: haskell-src/Concordium/Utils/BinarySearch.hs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+-- |Perform a binary search on a sorted vector, returning the index and the value, if found.
+-- The first argument is a function that resolves an entry in a monad.
+-- The second argument is a function that maps a value to its search key.
+-- The elements of the vector must be sorted in ascending order with respect to the key, with no
+-- duplicate keys.
+binarySearchIM ::
+    (Ord k, Monad m) =>
+    (b -> m a) ->
+    (a -> k) ->
```
Comment (abizjak) [inline haskell-src/Concordium/Utils/BinarySearch.hs:43]: Why two arguments? It does not look like the function first caches the results of applying the first function.

File: haskell-src/Concordium/Utils/BinarySearch.hs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    Vec.Vector b ->
+    k ->
+    m (Maybe (Int, a))
```
Comment (abizjak) [inline haskell-src/Concordium/Utils/BinarySearch.hs:46]: This is not really entirely specified due to the monadic first argument. The order in which elements of the vector are accessed should be specified, and further preconditions stated, such as the key that is retrieved is always "equivalent" no matter the state of the computation.

File: haskell-tests/Generators.hs [hunk 4]
```diff
@@ -0,0 +1,977 @@
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Generators where
```
Comment (abizjak) [inline haskell-tests/Generators.hs:7]: Provide module-level documentation as to why this exists.

File: haskell-tests/Generators.hs [hunk 5]
```diff
@@ -0,0 +1,977 @@
+{-# OPTIONS_GHC -fno-warn-orphans #-}
```
Comment (abizjak) [inline haskell-tests/Generators.hs:1]: It was unnecessary to have this refactoring now.

Either it should have been done sooner, in a separate PR, or later, in a different PR.

This is just adding needless complexity to the review.

File: haskell-tests/Generators.hs [hunk 6]
```diff
@@ -0,0 +1,977 @@
...
+            sLen <- choose (50, 70)
+            sig <- Signature . BSS.pack <$> vector sLen
+            return (idx, sig)
+        (,Map.fromList credentialSignatures) . CredentialIndex <$> arbitrary
+
+    let atrSignature = TransactionSignature (Map.fromList allKeys)
+    return $! makeAccountTransaction atrSignature atrHeader atrPayload
+
+baseTime :: UTCTime
+baseTime = read "2019-09-23 13:27:13.257285424 UTC"
```
Comment (abizjak) [inline haskell-tests/Generators.hs:715]: Documentation.

File: haskell-tests/Generators.hs [hunk 7]
```diff
@@ -0,0 +1,977 @@
...
+    kps <- vectorOf (fromIntegral thr * 2) genSigSchemeKeyPair
+    let hlkKeys = Vec.fromList $ correspondingVerifyKey <$> kps
+    return (HigherLevelKeys{hlkThreshold = thr, ..}, kps)
+
+genKeyCollection :: IsChainParametersVersion cpv => UpdateKeysThreshold -> Gen (UpdateKeysCollection cpv, [KeyPair], [KeyPair], [KeyPair])
+genKeyCollection thr = do
+    (rootKeys, a) <- genRootKeys thr
+    (level1Keys, b) <- genLevel1Keys thr
+    (level2Keys, c) <- genAuthorizationsAndKeys thr
+    return (UpdateKeysCollection{..}, a, b, c)
```
Comment (abizjak) [inline haskell-tests/Generators.hs:977]: This is a poor quality module. If refactoring was desired, then documentation should accompany it. As it is, it is a massive module with some generators, no clarity on the quality of these generators, or their coverage. 

While that was also true before this move, for refactoring to have value it should improve maintainability of code. This does not achieve that.

File: haskell-tests/Genesis/ParametersSpec.hs [hunk 8]
```diff
@@ -0,0 +1,44 @@
+{-# LANGUAGE DataKinds #-}
+
+module Genesis.ParametersSpec (tests) where
```
Comment (abizjak) [inline haskell-tests/Genesis/ParametersSpec.hs:3]: Provide a summary of what this is testing. Looks like just JSON serialization, but the module name implies a more general test.

File: haskell-tests/Types/AmountFraction.hs [hunk 9]
```diff
@@ -0,0 +1,35 @@
+{-# LANGUAGE OverloadedStrings #-}
+module Types.AmountFraction where
```
Comment (abizjak) [inline haskell-tests/Types/AmountFraction.hs:2]: Provide module level documentation as to what is the purpose of this test.

File: haskell-tests/Types/ParametersSpec.hs [hunk 10]
```diff
@@ -0,0 +1,63 @@
+{-# LANGUAGE DataKinds #-}
```
Comment (abizjak) [inline haskell-tests/Types/ParametersSpec.hs:1]: Module level documentation summarizing what is the purpose of this test.

File: haskell-tests/Types/PayloadSerializationSpec.hs [hunk 11]
```diff
@@ -7,256 +7,31 @@ import Test.QuickCheck.Monadic
...
-  Just eatd <- run (makeEncryptedAmountTransferData globalContext (_elgamalPublicKey public) private agg amount)
-  return (checkPayload SP1 (EncryptedAmountTransfer addr eatd))
+testSerializeEncryptedTransfer =
+  property $ \gen gen1 seed1 seed2 -> forAll genAccountAddress $ \addr -> monadicIO $ do
+    let public = AccountEncryptionKey . deriveElgamalPublicKey globalContext . generateGroupElementFromSeed globalContext $ seed1
+    let private = generateElgamalSecretKeyFromSeed globalContext seed2
+    let agg = makeAggregatedDecryptedAmount (encryptAmountZeroRandomness globalContext gen) gen (EncryptedAmountAggIndex gen1)
+    let amount = gen `div` 2
+    Just eatd <- run (makeEncryptedAmountTransferData globalContext (_elgamalPublicKey public) private agg amount)
+    return (checkPayload SP1 (EncryptedAmountTransfer addr eatd))
```
Comment (abizjak) [inline haskell-tests/Types/PayloadSerializationSpec.hs:34]: Please refrain from needless formatting. there is no need to spend time reviewing whether anything has changed or not.

File: haskell-tests/Types/PayloadSerializationSpec.hs [hunk 12]
```diff
@@ -278,21 +53,63 @@ groupIntoSize s =
...
 checkPayload :: SProtocolVersion pv -> Payload -> Property
 checkPayload spv e = let bs = S.runPut $ putPayload e
-                     in case S.runGet (getPayload spv (fromIntegral (BS.length bs))) bs of
-                          Left err -> counterexample err False
-                          Right e' -> label (groupIntoSize (fromIntegral (BS.length bs))) $ e === e'
+                 in case S.runGet (getPayload spv (fromIntegral (BS.length bs))) bs of
+                      Left err -> counterexample err False
+                      Right e' -> label (groupIntoSize (fromIntegral (BS.length bs))) $ e === e'
+
+modifyPayloadBitmap :: (Word16 -> Word16) -> BS.ByteString -> BS.ByteString
```
Comment (abizjak) [inline haskell-tests/Types/PayloadSerializationSpec.hs:60]: Provide documentation. The name is very generic and does not inform the reader about what this does.

File: haskell-bins/generate-update-keys/Main.hs [hunk 13]
```diff
@@ -45,40 +46,52 @@ readHigherAuthDetails = maybeReader $ \s -> case reads s of
     _ -> Nothing
   _ -> Nothing
 
+data CommonUpdateKeys = CommonUpdateKeys {
```
[the remark this answers is not in the document]
Comment (td202) [reply haskell-bins/generate-update-keys/Main.hs:50]: Added documentation.

File: haskell-bins/generate-update-keys/Main.hs [hunk 14]
```diff
@@ -115,30 +135,63 @@ main = customExecParser p opts >>= generateKeys
...
+        asParamMicroGTUPerEuro <- makeAS cukGTUEuro "GTU-Euro rate update access structure"
+        asParamFoundationAccount <- makeAS cukFoundationAccount "Foundation account update access structure"
+        asParamMintDistribution <- makeAS cukMintDistribution "Mint distribution update access structure"
+        asParamTransactionFeeDistribution <- makeAS cukTransactionFeeDistribution "Transaction fee distribution update access structure"
+        asParamGASRewards <- makeAS cukGASRewards "GAS rewards update access structure"
+        asBakerStakeThreshold <- makeAS cukBakerStakeThreshold "Baker minimum threshold access structure"
+        asAddAnonymityRevoker <- makeAS cukAddAnonymityRevoker "Add anonymity revoker access structure"
+        asAddIdentityProvider <- makeAS cukAddIdentityProvider "Add identity provider access structure"
+        cooldownParameters <- makeAS gukCooldownParameters "Add identity provider access structure"
+        timeParameters <- makeAS gukCooldownParameters "Add identity provider access structure"
```
[the remark this answers is not in the document]
Comment (td202) [reply haskell-bins/generate-update-keys/Main.hs]: Refactored

File: haskell-src/Concordium/Cost.hs [hunk 15]
```diff
@@ -94,6 +94,14 @@ scheduledTransferCost n = fromIntegral n * (300 + 64)
 addBakerCost :: Energy
 addBakerCost = 4050
 
+-- |C_t for configure baker when keys are not present
+configureBakerCostWithoutKeys :: Energy
+configureBakerCostWithoutKeys = 300
```
[the remark this answers is not in the document]
Comment (td202) [reply haskell-src/Concordium/Cost.hs:99]: The cost was set in line with `removeBakerCost`, as the process is pretty similar. More things can be changed as a result, but these are somewhat accounted for by the size of the transaction and resulting cost.

Is there a standard way we have of analysing transaction costs?

**Review by td202 — COMMENTED** [review]
I have attempted to address most of the concerns. Most of the changes to the API (beyond adding protocol version) I have not done. Since they cut across projects, I think such changes should be handled in a separate PR in order not to balloon the scope of the present one.

File: haskell-src/Concordium/Cost.hs (hunk 15 above)
[the remark this answers is not in the document]
Comment (abizjak) [reply haskell-src/Concordium/Cost.hs:99]: Yes, the costs assigned at present for transactions which have no other natural bound by setting up a local network and sending as many as possible, observing when the chain can no longer handle it.

For some costs where we can measure in a more isolated way (e.g., signature checking) we have benchmarks just for that. But these "per transaction costs" are as I outlined above.

**Review by abizjak — APPROVED** [review]

**Pull request merged**